De-experimentalize st.dialog

raethlein opened the pull request:
## Describe your changes

Merge **after** https://github.com/streamlit/streamlit/pull/9019

With https://github.com/streamlit/streamlit/pull/9019 landing, we are also de-experimentializing the `@st.experimental_dialog` decorator and introduce `@st.dialog`. As a consequence of the aforementioned PR, 
dialogs can then
- call other fragments or be opened from fragments, although dialogs are not allowed to open other dialogs
- dialogs can be opened by callbacks (closes https://github.com/streamlit/streamlit/issues/8591)

Update all used instances of `@st.experimental_dialog` to `@st.dialog`.

## GitHub Issue Link (if applicable)

## Testing Plan

- Unit Tests (JS and/or Python)
  - Added a unit test to ensure that the experimental-decorator still continues to work

Closes #9063

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

vdonato committed:
Don't immediately remove fragments added by callbacks (#8916)

The reason we were seeing the bug in #8591 is due to the incorrect assumption that we can always
totally clear out a session's fragment storage before a full script run as fragments will immediately
be repopulated anyway. This isn't quite true as a fragment can be added by a callback, and in this
case clearing out `FragmentStorage` before the full script run will lose it.

To fix this, we keep track of the fragments added throughout a script run, and on a full script run
we don't clear out fragments that were just added.

vdonato committed:
Remove fragment_ids_this_run from script run context (#8953)

I realized while working on the `st.rerun(scope="fragment")` feature that `ctx. fragment_ids_this_run`
should always be `rerun_data.fragment_id_queue` in set form. We still want to keep the distinction
around since the frontend isn't aware of the order fragments will be run, but we should simplify things
to not needlessly copy around the same information in a slightly different representation.

vdonato committed:
Implement st.rerun(scope="fragment") (#8979)

This PR adds the new `scope` kwarg to the `st.rerun()` function. With `scope="app"` (the default),
`st.rerun()` work as before. If you set `scope="fragment"` within a fragment, however, it's now possible
to specify that only the currently running fragment reruns.

vdonato committed:
Throw/expect more specific error when hitting KeyErrors in FragmentStorage methods (#9011)

A source of many (but potentially not all) of the reports in #8494 involve a silly mistake that I made
where we're catching all `KeyError`s that may be thrown from a fragment and interpreting them as
coming from when we attempted to fetch the fragment from our `FragmentStorage`. Of course, they
could also be coming from the fragment function itself, and in these cases we're currently raising an
error with a misleading / incorrect / opaque message.

This PR fixes this issue by throwing and expecting a more specific subclass of `KeyError` in
`FragmentStorage` operations

vdonato committed:
Remove mistakenly added whitespace

raethlein committed:
De-experimentalize fragment

raethlein committed:
Update e2e test

vdonato commented on lib/tests/streamlit/elements/layouts_test.py:
```diff
@@ -112,6 +114,7 @@ def test_not_equal_width_float_columns(self):
 
         for column in columns:
             with column:
+                """Noop"""
```
nit (here and throughout this test): could this be a normal comment rather than a docstring?

raethlein replied:
yep, the linter seems to accept that! I have updated it

vdonato commented on lib/streamlit/elements/dialog_decorator.py:
```diff
@@ -133,8 +142,8 @@ def dialog_decorator(
         fragments, and fragments can't contain dialogs. Using dialogs in widget
         callback functions is not supported.
```
nit: need to update this docstring to remove the limitations about dialogs not being able to call fragments + not being trigger-able by widget callbacks

raethlein replied:
oh yeah great catch, thanks!

raethlein committed:
Address feedback

raethlein committed:
Always add fragment_ids to set (#9069)

Fix two issues:
- the `current_fragment_delta_path` could have been taken from the wrong
snapshot
- always add the `current_fragment_id` to the set of new fragment ids.
This prevents an issue where a fragment rerun is triggered during full
app execution.

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Add e2e test to ensure that the `current_fragment_id` error does not
happen
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

raethlein committed:
Fix duplicate dialog fragment id (#9066)

Right now, dialogs within an app always generate the same fragment id.
This is problematic and can lead to mixing contents of different
dialogs. Add a new parameter that is used by the dialog decorator to
pass the actually wrapped function name to the fragment in order to
generate the hash.
This behavior started to surface in the `feature/st.fragment` branch; prior to that we overwrote
fragments with the same id in the storage and since we only allow a single dialog at any given time, this issue stayed unrecognized.

- Unit Tests (JS and/or Python)
- Unit tests seem difficult to add for this because if we use the
`DeltaGeneratorTestCase`, the dialogs have different delta paths when
opened simultaneously versus the same delta path in a real app after a
rerun happened. Hence, I figured the e2e test resembles the chain of
sequence best.
- E2E Tests
- Add an e2e test that triggers the sequence of action that leads to the
bug

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.

raethlein committed:
Fix fragment exception impacts queued fragments (#9041)

In the current state, an exception in a fragment will cause queued
fragments to stop from running. But instead we want queued fragments to
still run even if a previous fragment raised an exception (even if the
error happens in a full app run; the error will be displayed but the app
will continue to run).

To do so, we add except blocks into the loop and continue running the
other fragments in the queue.

Also, we want to stop app execution when a fragment raises an error
during a full app run. This is why the `exec_func_with_error_handling`
call is removed in `fragment.py`.

- Explanation of why no additional tests are needed
- Unit Tests (JS and/or Python)
- E2E Tests
- Any manual testing needed?

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this
project are made under the Apache 2.0 license.
